File: go/tree-building/tree_building.go
// Package tree builds trees.
package tree

import (
	"errors"
	"sort"
)

// Record stores a DB record.
type Record struct {
	ID     int
	Parent int
}

// Node represents a tree node.
type Node struct {
	ID       int
	Children []*Node
}

// Build generates a tree of nodes from a list of records.
func Build(records []Record) (*Node, error) {
	nodes := make(map[int]*Node)
	children := make(map[int]sort.IntSlice)

	for _, r := range records {
		if _, ok := nodes[r.ID]; ok {
			return nil, errors.New("node defined twice")
		}
		if r.ID < r.Parent || (r.ID == r.Parent && r.ID != 0) {
			return nil, errors.New("parent ID must be smaller than child")
		}
		if r.ID == 0 && r.Parent != 0 {
			return nil, errors.New("root node cannot have a parent")
		}
		nodes[r.ID] = &Node{r.ID, nil}
		children[r.Parent] = append(children[r.Parent], r.ID)
	}

	for i := 0; i < len(children); i++ {
		if _, ok := children[i]; !ok {
			return nil, errors.New("non-continuous IDs")
		}
	}

	for p, cs := range children {
		p, ok := nodes[p]
		if !ok {
			return nil, errors.New("invalid parent referenced")
		}
		cs.Sort()
		for _, c := range cs {
			if c == 0 {
				continue
			}
			p.Children = append(p.Children, nodes[c])
		}
	}
	return nodes[0], nil
}

File: python/tree-building/tree_building.py
import dataclasses


@dataclasses.dataclass
class Record:
    record_id: int
    parent_id: int


@dataclasses.dataclass
class Node:
    node_id: int
    def __post_init__(self) -> None:
        self.children: list[Node] = []


def BuildTree(records: list[Record]) -> Node | None:
    if not records:
        return None

    nodes = {}
    records.sort(key=lambda x: x.record_id)

    for record in records:
        if record.record_id != 0 and record.record_id == record.parent_id:
            raise ValueError('Only root should have equal record and parent id.')
        if record.record_id < record.parent_id:
            raise ValueError("Node parent_id should be smaller than it's record_id.")

        nodes[record.record_id] = Node(record.record_id)

    if 0 not in nodes:
        raise ValueError('Record id is invalid or out of order.')
    if sorted(nodes.keys()) != list(range(len(nodes))):
        raise ValueError('Record id is invalid or out of order.')

    for record in records:
        if record.record_id == 0:
            continue
        nodes[record.parent_id].children.append(nodes[record.record_id])
    
    return nodes[0]
